File: the-advanced-webservice/typescript/lib/the-advanced-webservice-construct.ts
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as route53 from 'aws-cdk-lib/aws-route53';
import * as targets from 'aws-cdk-lib/aws-route53-targets';
import * as rds from 'aws-cdk-lib/aws-rds';
import * as secret from 'aws-cdk-lib/aws-secretsmanager';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { HttpUrlIntegration, HttpLambdaIntegration } from 'aws-cdk-lib/aws-apigatewayv2-integrations';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import * as origins from 'aws-cdk-lib/aws-cloudfront-origins';
import * as certificate from 'aws-cdk-lib/aws-certificatemanager';
import { App, Duration, Stack, NestedStack, NestedStackProps, StackProps } from 'aws-cdk-lib';
import * as logs from 'aws-cdk-lib/aws-logs';


export interface TheAdvancedWebserviceConstructProps extends StackProps {
  //define construct properties here
  //required properties
  createvpc?: boolean;
  createlambda?: boolean;
  createrdscluster?: boolean;
  createhostedzone?: boolean;
  apiName?: string;
  stageName?: string;
  originPath?: string;

  //define the below additional properties when createvpc is set to true
  vpcAz?: number;
 
  //define the below additional property when using an existing vpc and createvpc is set to false
  vpcId?: string;

  //define the below additional property if createhostedzone is set to true
  createpublichostedzone?: boolean;

  //define this additional property if createlambda is set to false
  functionArn?: string;

  //define this additional property if createlambda is set to true
  lambdaCodeFilePath?: string; //the file directory where lambda zipped file is
  runTime?: lambda.Runtime |any; //the lambda runtime
  

  //define this additional property if createrds is set to false
  existingRdsClusterIdentifier?: string;

  //define this additional properties if createrds is set to true
  databaseEngine?: rds.DatabaseClusterEngine |any;

  //define this additional property if createhostedzone is set to false
  existingDomainName?: string; //TODO this fails to compile when not provided...
  existingHostedZoneId?: string;

  //define this additional property if createhostedzone is set to true 
  newDomainName?: string;
  recordSubDomainName?: string;
  
}

//Define L3 Construct
export class TheAdvancedWebserviceConstruct extends Construct {

  constructor(scope: Construct, id: string, props: TheAdvancedWebserviceConstructProps = {}) {
    super(scope, id);

                    // CONSTRUCT CONTENTS

  //DEFINE VPC RESOURCES

  //condition to allow use of either an existing vpc or create a new one based on context values declared

    let myvpc;

    if (props.createvpc==false) {
      myvpc = ec2.Vpc.fromLookup(this, 'testvpc', { 
         vpcId: props.vpcId
    
      });     
    }
    else {
      myvpc = new ec2.Vpc(this, 'testvpc', {
         ipAddresses: ec2.IpAddresses.cidr('10.0.0.0/16'),
         maxAzs: props.vpcAz
      });
    }
 
  //DEFINE LAMBDA RESOURCES

  //declare context values

  const existingfunctionname = this.node.tryGetContext('existingfunctionname');

//condition to allow use of either an existing aurora cluster or create a new aurora cluster based on context value

  let mylambda;
  let mylambdarole;

  let lambdacodefilepath:string = props.lambdaCodeFilePath!;

  let functionarn:string = props.functionArn!;

  if (props.createlambda==false) {
    mylambda = lambda.Function.fromFunctionAttributes(this, 'mylambda', {
      sameEnvironment: true,
      functionArn: functionarn
    });     
  }
  else {
    mylambdarole = new iam.Role(this, 'mylambdaexecutionrole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      roleName: 'mylambdaexecutionrole',
      managedPolicies: [
        iam.ManagedPolicy.fromManagedPolicyArn(this, 'mypolicy','arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole'), 
        iam.ManagedPolicy.fromManagedPolicyArn(this, 'mypolicy2', 'arn:aws:iam::aws:policy/AmazonEC2FullAccess'),
      ]
    });


    mylambda = new lambda.Function(this, 'mylambda', {
      runtime: props.runTime,
      code: lambda.Code.fromAsset(lambdacodefilepath),
      handler: 'rdsLambda.handler',
      vpc: myvpc,
      role: mylambdarole,
      vpcSubnets: {subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,},

    });
  }

  //DEFINE RDS RESOURCES

  //condition to allow use of either an existing aurora cluster or create a new aurora cluster based on context values declared

    let myrds;
    let myclustersubnetgroup;
    let myclusterparametergroup;    
    let myclustercredentials;
    let myclusterproxy;
    let existingrdsclusteridentifier:string = props.existingRdsClusterIdentifier!; 

    if (props.createrdscluster==false) {
      myrds = rds.DatabaseCluster.fromDatabaseClusterAttributes(this, 'testrds',{
        clusterIdentifier: existingrdsclusteridentifier,
      });
    }

    else {
      myclustersubnetgroup = new rds.SubnetGroup(this, 'testclustersubnetgroup',{
        vpc: myvpc,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        description: 'test subnet group',
        vpcSubnets: {
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
        }
      } );

      myclusterparametergroup = new rds.ParameterGroup(this, 'testclusterparametergroup',{
        engine: props.databaseEngine,
        parameters: {
          aws_default_lambda_role: 'mylambdaexecutionrole',
        }
        
      });

      myclustercredentials =  new rds.DatabaseSecret(this, 'myrdsscreds', {
        username: 'ClusterAdmin',
        secretName: 'myrdsscreds',

      });

      myrds = new rds.DatabaseCluster(this, 'testauroracluster', {
         engine: props.databaseEngine,         
         vpc: myvpc,
         defaultDatabaseName: 'demos',
         subnetGroup: myclustersubnetgroup,
         parameterGroup: myclusterparametergroup,
         credentials: { username: 'clusteradmin' },
         serverlessV2MinCapacity: 2,
         serverlessV2MaxCapacity: 64,
         port: 3306,
         writer: rds.ClusterInstance.serverlessV2 ('writer'),
         readers: [rds.ClusterInstance.serverlessV2 ('reader', {scaleWithWriter: true})],
         storageEncrypted: true,
         storageType: rds.DBClusterStorageType.AURORA,

      });

      myrds.addRotationSingleUser({
        automaticallyAfter: cdk.Duration.days(30),
        rotateImmediatelyOnUpdate: true,
        
      });

      //add proxy and lambda connection via the proxy to the rds database cluster
      myclusterproxy = new rds.DatabaseProxy(this, 'rdsProxy', {
        proxyTarget: rds.ProxyTarget.fromCluster(myrds),
        secrets: [myclustercredentials],
        vpc: myvpc
      });

      myclusterproxy.grantConnect( iam.Role.fromRoleName(this, 'mylambdarole','mylambdaexecutionrole'))

    }

  //ensure the rds is not created before the lambda function 

    myrds.node.addDependency(mylambda)
 
  //DEFINE API GATEWAY RESOURCES

  //this resource requires creation of a new resource

  //define api resources
  let apiname:string = props.apiName!; 

  let stagename:string = props.stageName!; 

  const DefaultIntegration = new apigateway.LambdaIntegration(mylambda);

  const apiLogGroup = new logs.LogGroup(this, "apiLogs");

  let myapi = new apigateway.RestApi(this, 'myapi', { 
    restApiName: apiname
  
  });

  const deployment = new apigateway.Deployment(this, 'Deployment', {
    api: myapi,
    retainDeployments: true
  
  });

  const stage = new apigateway.Stage(this, 'Stage', {
    deployment: deployment,
    stageName: stagename,
    accessLogDestination: new apigateway.LogGroupLogDestination(apiLogGroup),
    accessLogFormat: apigateway.AccessLogFormat.jsonWithStandardFields(),
    tracingEnabled: true,
    loggingLevel: apigateway.MethodLoggingLevel.INFO
  })

  const method = new apigateway.Method(this, 'Method', {
    httpMethod: 'GET',
    resource: myapi.root,
    integration: DefaultIntegration 
  })

  //DEFINE CLOUDFRONT RESOURCES

  let originpath:string = props.originPath!; 

  //define resources

  const mydistribution = new cloudfront.Distribution(this, 'mydist', {
    defaultBehavior: {
      origin: new origins.RestApiOrigin(myapi,{originPath: originpath}),
      allowedMethods: cloudfront.AllowedMethods.ALLOW_ALL,
      viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
      
    },
    minimumProtocolVersion: cloudfront.SecurityPolicyProtocol.TLS_V1_2_2021 
  })

  //ensure the distribution is not created before the api 

  mydistribution.node.addDependency(myapi)
 
  //DEFINE ROUTE 53 RESOURCES
 
//condition to allow use of either an existing hosted zone or create a new public or private hosted zone based on context values declared

  let myhostedzone;
  let existingdomainname:string = props.existingDomainName!;
  let existinghostedzoneid:string = props.existingHostedZoneId!;
  let newdomainname:string = props.newDomainName!;
  let recordsubdomainname:string = props.recordSubDomainName!;

  if (props.createhostedzone==false) {
    myhostedzone = route53.HostedZone.fromHostedZoneAttributes(this, 'testhostedzone', {
       zoneName: existingdomainname,
       hostedZoneId: existinghostedzoneid,

    });
  } else if (props.createpublichostedzone==false) {
    myhostedzone = new route53.PrivateHostedZone(this, 'testhostedzone', {
       zoneName: newdomainname,
       vpc: myvpc
    });
  } else {
    myhostedzone = new route53.PublicHostedZone(this, 'testhostedzone', {
       zoneName: newdomainname
    });
  }

  const myrecordset = new route53.ARecord(this, 'myrecord', {
    zone: myhostedzone,
    target: route53.RecordTarget.fromAlias( new targets.CloudFrontTarget(mydistribution)),
    recordName: recordsubdomainname
  })

  //ensure the recordset is not created before the distribution

  myrecordset.node.addDependency(mydistribution)
 
 
  }
}
 

File: the-advanced-webservice/python/the_advanced_webservice/the_advanced_webservice_construct.py
from aws_cdk import (
    Duration,
    App,
    Stack,
    Tags,
    aws_lambda as lambda_,
    aws_ec2 as ec2,
    aws_cloudfront as cloudfront,
    RemovalPolicy,
    aws_apigateway as apigw,
    aws_logs as logs,
    aws_route53_targets as targets,
    aws_route53 as route53,
    aws_cloudfront_origins as origins,
    aws_rds as rds,
)
from constructs import Construct
from aws_cdk import aws_lambda as lambda_
from aws_cdk import aws_iam as iam


class TheAdvancedWebservice(Construct):

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        create_vpc: bool,
        vpc_az: int,
        create_lambda: bool,
        create_hostedzone: bool,
        create_rdscluster: bool,
        lambda_code_file_path: str,
        run_time: lambda_.Runtime,
        database_engine: rds.IClusterEngine,
        api_name: str,
        stage_name: str,
        origin_path: str,
        domain_name: str,
        existing_hosted_zone_id: str,
        vpc_id: str = None,
        function_arn: str = None,
        # existing_domain_name: str = None,
        # new_domain_name: str = None,
        # existing_hosted_zone_id: str = None,
        existing_rds_cluster_identifier: str = None,
        record_subdomain_name: str = None,
        create_public_hosted_zone: bool = False,
        **kwargs
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        myvpc = None

        if create_vpc:
            myvpc = ec2.Vpc(
                self,
                "NewVpc",
                max_azs=vpc_az,
                ip_addresses=ec2.IpAddresses.cidr("10.0.0.0/16"),
            )
        else:
            myvpc = ec2.Vpc.from_lookup(self, "ImportedVpc", vpc_id=vpc_id)

        lambda_function = None
        lambda_role = None

        if create_lambda:
            lambda_role = iam.Role(
                self,
                "LambdaExecutionRole",
                assumed_by=iam.ServicePrincipal("lambda.amazonaws.com"),
                managed_policies=[
                    iam.ManagedPolicy.from_managed_policy_arn(
                        self,
                        "AWSLambdaBasicExecutionRole",
                        "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
                    ),
                    iam.ManagedPolicy.from_managed_policy_arn(
                        self,
                        "AmazonEC2FullAccess",
                        "arn:aws:iam::aws:policy/service-role/AmazonEC2FullAccess",
                    ),
                ],
            )

            lambda_function = lambda_.Function(
                self,
                "NewLambda",
                runtime=run_time,
                code=lambda_.Code.from_asset(lambda_code_file_path),
                handler="rdsLambda.handler",
                vpc=myvpc,
                role=lambda_role,
            )
        else:
            lambda_function = lambda_.Function.from_function_attributes(
                self,
                "ImportedLambda",
                function_arn=function_arn,
                same_environment=True,
            )

        myrds = (None,)

        if create_rdscluster:
            subnet_group = rds.SubnetGroup(
                self,
                "MySubnetGroup",
                description="test subnet group",
                vpc=myvpc,
                removal_policy=RemovalPolicy.DESTROY,
                subnet_group_name="subnetGroupName",
                vpc_subnets=ec2.SubnetSelection(
                    subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS,
                ),
            )

            parameter_group = rds.ParameterGroup(
                self,
                "ParameterGroup",
                engine=database_engine,
                parameters={"aws_default_lambda_role": "mylambdaexecutionrole"},
            )

            my_cluster_credentials = rds.DatabaseSecret(
                self, "myrdsscreds", username="ClusterAdmin", secret_name="myrdsscreds"
            )

            myrds = rds.DatabaseCluster(
                self,
                "testauroracluster",
                engine=database_engine,
                default_database_name="demos",
                credentials=rds.Credentials.from_secret(
                    my_cluster_credentials, username="clusteradmin"
                ),
                parameter_group=parameter_group,
                subnet_group=subnet_group,
                storage_encrypted=True,
                port=3306,
                writer=rds.ClusterInstance.serverless_v2("writer"),
                readers=[
                    rds.ClusterInstance.serverless_v2("reader", scale_with_writer=True),
                ],
                serverless_v2_max_capacity=64,
                serverless_v2_min_capacity=2,
                vpc=myvpc,
                storage_type=rds.DBClusterStorageType.AURORA,
            )

            myrds.add_rotation_single_user(
                automatically_after=Duration.days(30),
                rotate_immediately_on_update=True,
            )

            my_cluster_proxy = rds.DatabaseProxy(
                self,
                "myrdsproxy",
                secrets=[my_cluster_credentials],
                proxy_target=rds.ProxyTarget.from_cluster(myrds),
                vpc=myvpc,
            )

            my_cluster_proxy.grant_connect(
                iam.Role.from_role_name(
                    self,
                    "mylambdarole",
                    role_name="mylambdaexecutionrole",
                )
            )

        else:
            myrds = rds.DatabaseCluster.from_database_cluster_attributes(
                self,
                "testrds",
                cluster_identifier=existing_rds_cluster_identifier,
            )

        myrds.node.add_dependency(lambda_function)

        my_api = apigw.RestApi(
            self,
            "myapi",
            rest_api_name=api_name,
        )

        deployment = apigw.Deployment(
            self,
            "Deployment",
            api=my_api,
            retain_deployments=True,
        )

        api_log_group = logs.LogGroup(
            self,
            "apiLogs",
        )
        stage = apigw.Stage(
            self,
            "Stage",
            deployment=deployment,
            stage_name=stage_name,
            access_log_destination=apigw.LogGroupLogDestination(api_log_group),
            access_log_format=apigw.AccessLogFormat.clf(),
            tracing_enabled=True,
            logging_level=apigw.MethodLoggingLevel.INFO,
        )

        method = apigw.Method(
            self,
            "Method",
            http_method="GET",
            resource=my_api.root,
            integration=apigw.LambdaIntegration(lambda_function),
        )

        my_distro = cloudfront.Distribution(
            self,
            "mydist",
            default_behavior=cloudfront.BehaviorOptions(
                origin=origins.RestApiOrigin(my_api, origin_path=origin_path)
            ),
        )

        if create_hostedzone is False and existing_hosted_zone_id is not None:
            my_hosted_zone = route53.HostedZone.from_hosted_zone_attributes(
                self,
                "testhostedzone",
                hosted_zone_id=existing_hosted_zone_id,
                zone_name=domain_name,
            )
        elif create_public_hosted_zone is False:
            my_hosted_zone = route53.PrivateHostedZone(
                self,
                "testhostedzone",
                zone_name=domain_name,
                vpc=myvpc,
            )
        else:
            my_hosted_zone = route53.PublicHostedZone(
                self,
                "testhostedzone",
                zone_name=domain_name,
            )

        my_record_set = route53.ARecord(
            self,
            "myrecord",
            zone=my_hosted_zone,
            record_name=record_subdomain_name,
            target=route53.RecordTarget.from_alias(targets.CloudFrontTarget(my_distro)),
        )
        my_record_set.node.add_dependency(my_distro)
